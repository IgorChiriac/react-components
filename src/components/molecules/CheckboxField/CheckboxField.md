### Summary

`<CheckboxField />` is a convenience wrapper around HTML `<input type="checkbox" />` and two **atoms**:

- [`<InputLabel />`](/#/Components/Atoms/InputLabel)
- [`<ErrorMessage />`](/#/Components/Atoms/ErrorMessage)

⚠️ &nbsp;Note that a `name` prop **must be provided** to automatically set:

- `htmlFor` prop in `<InputLabel />`
- `data-automation` prop in `<ErrorMessage />`
- `id` prop in `<input type="checkbox" />` for accessibility and automation purposes

### Examples

- Default

```tsx
import { CheckboxField } from '@zopauk/react-components';

<CheckboxField label="Do you love jazz?" name="check1" />;
```

- Checked state

```tsx
import { CheckboxField } from '@zopauk/react-components';

<CheckboxField label="Do you love rap?" name="text2" defaultChecked={true} />;
```

- With an error

```tsx
import { CheckboxField } from '@zopauk/react-components';

<CheckboxField label="Do you agree?" errorMessage="Oops! You forgot to check this" name="text6" />;
```
