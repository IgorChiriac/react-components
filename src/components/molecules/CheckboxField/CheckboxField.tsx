import React from 'react';
import styled, { keyframes } from 'styled-components';
import { colors, typography } from '../../../constants';
import tealCheckMark from '../../../content/images/teal-check-mark.svg';
import greyCheckMark from '../../../content/images/grey-check-mark.svg';
import ErrorMessage from '../../atoms/ErrorMessage/ErrorMessage';
import InputLabel from '../../atoms/InputLabel/InputLabel';
import SizedContainer from '../../layout/SizedContainer/SizedContainer';
import { getBorderColorByStatus } from '../../../helpers/utils';
import { FieldProps, InputProps } from '../../types';

export interface CheckboxFieldProps extends FieldProps, InputProps {
  name: string;
}

const zoomOut = keyframes`
  from {
    transform: scale(0.4);
  }
  to {
    transform: scale(1);
  }
`;

const Input = styled.input<InputProps>`
  left: -100%;
  opacity: 0;
  z-index: -1;
  position: absolute;
  & + label {
    &:before {
      border-color: ${getBorderColorByStatus};
    }
  }
  &:checked + label {
    &:before {
      border-color: ${colors.brand};
    }
    &:after {
      background-size: contain;
      background-image: ${`url(${tealCheckMark})`};
      animation: ${zoomOut} 180ms ease-in-out;
    }
  }
  &:focus + label {
    &:before {
      border-color: ${colors.brand};
      box-shadow: 0 0 4px 0 ${colors.brand};
    }
  }
  &:disabled + label {
    cursor: not-allowed;
    color: ${colors.grey};
    &:before {
      border-color: ${colors.greyLight};
    }
  }
  &:disabled:checked + label {
    &:after {
      background-size: contain;
      background-image: ${`url(${greyCheckMark})`};
      animation: ${zoomOut} 180ms ease-in-out;
    }
  }
`;

const Label = styled(InputLabel)`
  width: auto;
  display: flex;
  align-items: center;
  font-weight: ${typography.weights.regular};
  font-size: ${typography.sizes.text.body};
  line-height: 1.4;
  color: ${colors.greyDarkest};
  position: relative;
  user-select: none;
  margin-bottom: 0;

  &:before {
    content: '';
    background-color: ${colors.white};
    border-radius: 6px;
    height: 24px;
    width: 24px;
    flex-shrink: 0;
    margin-right: 8px;
    transition-property: border, box-shadow;
    transition: 0.2s ease-in-out;
    box-shadow: 0 0 4px 0 transparent;
    border: 1px ${colors.grey} solid;
    display: block;
  }
  &:after {
    content: '';
    background-size: contain;
    background: transparent none no-repeat center;
    height: 12px;
    width: 12px;
    position: absolute;
    left: 6px;
    top: 6px;
  }
  &:hover {
    cursor: pointer;
    &:before {
      border: 1px ${colors.brand} solid;
    }
  }
`;

const FieldError = styled(ErrorMessage)`
  margin-top: 5px;
`;

const CheckboxField = (props: CheckboxFieldProps) => {
  const { label, errorMessage, className, inputSize, name, ...rest } = props;

  return (
    <>
      <SizedContainer size={inputSize} className={className}>
        <Input id={`checkbox-id-${name}`} type="checkbox" name={name} {...rest} />
        <Label htmlFor={`checkbox-id-${name}`}>{label}</Label>
      </SizedContainer>
      {errorMessage && <FieldError>{errorMessage}</FieldError>}
    </>
  );
};

export default CheckboxField;
