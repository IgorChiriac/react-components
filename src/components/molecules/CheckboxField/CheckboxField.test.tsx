import { axe } from 'jest-axe';
import React from 'react';
import { render } from 'react-testing-library';
import CheckboxField from './CheckboxField';

describe('<CheckboxField />', () => {
  it('renders the component with props with no a11y violations', async () => {
    const { container } = render(<CheckboxField label={'hello'} inputProps={{ name: 'test1' }} />);
    expect(container.firstChild).toMatchSnapshot();
    const results = await axe(container.innerHTML);
    expect(results).toHaveNoViolations();
  });
});
