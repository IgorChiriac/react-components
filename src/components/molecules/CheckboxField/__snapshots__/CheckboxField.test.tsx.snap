// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxField /> renders the component with props with no a11y violations 1`] = `
.c4 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #0D0A38;
  line-height: 1.6;
  font-family: "Open Sans",-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  font-weight: 600;
  font-size: 16px;
}

.c3 {
  display: block;
}

.c0 {
  max-width: 100%;
  width: 100%;
}

.c1 {
  left: -100%;
  opacity: 0;
  z-index: -1;
  position: absolute;
}

.c1:checked + label {
  font-weight: 600;
}

.c1:checked + label:before {
  content: '';
  background-color: #4A3EDE;
}

.c1:checked + label:after {
  content: '';
  background-size: contain;
  background-image: url(white-check-mark.svg);
  -webkit-animation: guYtTV 180ms ease-in-out;
  animation: guYtTV 180ms ease-in-out;
}

.c1:focus + label:before {
  border: 2px #4A3EDE solid;
}

.c2 {
  width: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: 400;
  color: #0D0A38;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2:before {
  content: '';
  background-color: #EAEAEE;
  border-radius: 6px;
  height: 24px;
  width: 24px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  margin-right: 8px;
  border: 2px #FFFFFF solid;
  display: block;
}

.c2:after {
  content: '';
  background-size: contain;
  background: transparent none no-repeat center;
  height: 12px;
  width: 12px;
  position: absolute;
  left: 6px;
  top: 6px;
}

.c2:hover {
  cursor: pointer;
}

.c2:hover:before {
  border: 2px #C5C6CD solid;
}

<div
  class="c0"
>
  <input
    class="c1"
    id="checkbox-id-test1"
    name="test1"
    type="checkbox"
  />
  <label
    class="c2 c3 c4"
    color="#0D0A38"
    for="checkbox-id-test1"
  >
    hello
  </label>
</div>
`;
